File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0547.py
class Solution:
    """
    Each time dfs finishes, we have traversed all nodes of a connected component
    """    
    BLACK = 0
    WHITE = 1
    
    def dfs(self, u, isConnected, color):
        color[u] = Solution.BLACK
        for v in range(len(isConnected[u])):
            if isConnected[u][v] and color[v] == Solution.WHITE:
                self.dfs(v, isConnected, color)    
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        color = [Solution.WHITE]*n
        
        noProvinces = 0
        for i in range(n):
            if color[i] == Solution.WHITE:
                noProvinces += 1
                self.dfs(i, isConnected, color)        
        return noProvinces
        

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0547.cpp
class Solution {
public:
    // Each time dfs finishes, we have traversed all nodes
    // of a connected component
    #define BLACK 0
    #define WHITE 1
    
    void dfs(int u, vector<vector<int>>& isConnected, vector<char> &color){
        color[u] = BLACK;
        for (int v=0; v<isConnected[u].size(); v++)
            if (isConnected[u][v] && color[v] == WHITE)
                dfs(v, isConnected, color);
    } // end-dfs
    
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();

        vector<char> color(n, WHITE);
        
        int noProvinces = 0;
        for (int i=0; i<n; i++){
            if (color[i] == WHITE){
                noProvinces++;
                dfs(i, isConnected, color);
            } // end-if
        } // end-for
        
        return noProvinces;
    } // end-findCircleNum
};